### PR TITLE
Issue#762 logs on demand

### PR DESCRIPTION
**Pull request by MFlyer**
PR on #762  Issue for Log utility - to be decided if to go on with this (like a version 1) and then update or wait till rotating logs integration

### CODE REVIEW

**Comment by MFlyer** [comment]
Hi @schakrava, code revision for #762 closed over 169b09f, PR ready for your revision before merging :smiley: 

Important remarks:
- Spawning has been added for all functions, allowing for example current log download while still reading it on a cat -n and obviously better processes management by gevent
- Cat and tail -n * calls always end (we assume that for true, cos' it can't be different), so we don't care about "safe killing", otherwise done for tail -f (it hangs forever till user action and we fire it on modal closing or on disconnect - accidental browser end/crash/etc)
- Error handling for empty or 0 size logs is transparent to user (no debug info on console/error raising, but direct info to user via log reader modal)
- I didn't forget about @phillxnet suggestion for systemd info (I totally agree about its usefulness! :wink: ), but imho it makes more sense under "Services" and not on logs
- If you think to move LogManagerNamespace to another file and import it...yes, do it! :+1: (that would divide widgets from LogManager & eventually future websockets)

Flyer/Mirko

File: src/rockstor/smart_manager/data_collector.py [hunk 1]
```diff
@@ -23,6 +46,265 @@
...
+            self.livereader_process.kill()
+            self.livereader_process = None
+            
+        #Emit current state only for checking purpose, no funcs handle on frontend
+        self.emit('logManager:logskilltailf', {
+            'key' : 'logManager:logskilltailf', 'data': {
+            'status' : self.livereading
+            }
+        })
+        gevent.sleep(0)
```
Comment (schakrava) [inline src/rockstor/smart_manager/data_collector.py]: Is gevent.sleep(0) necessary here? can you explain your rationale please?

File: src/rockstor/smart_manager/data_collector.py [hunk 2]
```diff
@@ -23,6 +46,265 @@
...
+                live_out = self.livereader_process.stdout.readline()
+                self.emit('logManager:logcontent', {
+                    'key': 'logManager:logcontent', 'data': {
+                    'current_rows' : 1,
+                    'total_rows' : 1,
+                    'chunk_content' : live_out,
+                    'content_size' : 1
+                    }
+                })
+                gevent.sleep(0)
```
Comment (schakrava) [inline src/rockstor/smart_manager/data_collector.py]: Can you please explain the rationale behind putting gevent.sleep(0) in the while loop here?

**Comment by schakrava** [comment]
Rotated logs section is appearing multiple times in the dropdown. Why is that?
screenshot from 2016-05-18 16 43 28

**Comment by schakrava** [comment]
When I click on more than one item in the Available Logs and download, the requested_logs.tgz file only has the log file corresponding to the first item. Is this a bug or am I using/expecting wrong things?

**Comment by schakrava** [comment]
My one word reaction to this pull request is an emphatic wow!!!

I couldn't have done it any better myself. I am really enjoying just playing with this feature, viewing and downloading various logs. I can see my browser memory usage go up and browser lock up because the data is being piped fast. I am sure we can make this a bit more efficient -- something to look into in near future. May be sleeping a bit more than the default 0 seconds could help?

Please take a look at all of my comments above and let me know your response. happy to test again with any new changes. @phillxnet and @priyaganti , you might want to deploy this branch, it's really fun to test!

**Comment by MFlyer** [comment]
Hi @schakrava , I'm pleased to see your reaction to this PR :smiley: 

Talking about gevent.sleeps : 
- the first one on the emit is not required (there's not a rational reason eheh)
- the second one on the live reading with tail -f : while we are live reading in the modal window we can't do anything else except for downloading the log file (link inside modal), so we don't wont our func to sleep showing us all stdout asa there's something to render, but you're right saying it makes no particular sense to gevent.sleep inside while loop! (if modal get closed loop get killed, so that gevent.sleep isn't necessary at all)

Talking about "bugs" on logs list and avail-downloadable logs : going to check on Firefox, cos' under chrome it was all fine

**Comment by MFlyer** [comment]
Found why you get duplicated elements on logs list under firefox:


While on Chrome (and Chromium too, i think) websockets stay alive, on firefox clicking the download link kills the websocket (see image); Rotated logs list is fetched on connection time on data_collector, so when the link get clicked our websocket dies, firefox reopens it and we get again rotated logs list...and so on every time you download a log file (both from archive builder and log download from modal window) :sob: 
websocketkilled

Going to investigate and find a workaround

> When I click on more than one item in the Available Logs and download, the requested_logs.tgz file only has the log file corresponding to the first item. Is this a bug or am I using/expecting wrong things?

We expect a tar file with all selected logs and this is what i get both on FF and Chrome

**Comment by MFlyer** [comment]
https://github.com/rockstor/rockstor-core/pull/1312#issuecomment-220191294 - @schakrava : solved with cc3af6a :wink: Tested both on chrome/firefox under win (I love and "hate" guys from Mozilla eheheh, but finally they're right saying "Ok, if you click a link you're leaving so you don't need that websocket anymore and we close it!")

**Comment by MFlyer** [comment]
Gevent.sleep(0) cleaning on 5a4b175 (those 2 referenced by @schakrava + others)

Last missing "bug" I'm not able to reproduce is requested_logs file having only one log: both on Chrome/Firefox tar file has all requested logs (opening with winrar under win).
Possible workaround: create requested_logs like requested_logs_UTC and check if it's just a kind of "file caching" <- @schakrava can you investigate if it's like this? :smiley: 

Flyer/Mirko

P.S. for fun: yesterday was sunny, tomorrow will be sunny, yesterday i was at work, tomorrow i'll be at work, today i'm on holiday (going to post office, get tickets for theater, etc etc)....today in Milan is rainy xD

**Comment by schakrava** [comment]
Thanks @MFlyer . I am unable to reproduce the problem either. May be i did not test properly yesterday...

Anyway, everything looks great to me! http://lgtm.in/i/Qegg5iK6q

**Comment by henfri** [comment]
Hello,
Wouldn't it make sense to optionally Filter the Logs with the help of logcheck?

**Comment by MFlyer** [comment]
Good morning @henfri,
Logs manager behaviour is to let user read logs / download them on demand, while [logcheck](http://linux.die.net/man/8/logcheck) acts in another way (mail root about log changes): do you agree on that?

Flyer

**Comment by henfri** [comment]
Yes and no.

Yes, logcheck mails log contents (not really changes, but triggered by time).
But the neat thing about logcheck is that it filters out the rubbish, noone is interested in. 
Thus, the conbination of your Userinterface and logchecks abilities to filter (frankly, it's just regexes that you can use directly in the userinterface without any of the logcheck binaries/scripts)

It is really great, because it helps uncovering issues and to focus.

**Comment by phillxnet** [comment]
@henfri and @MFlyer We do have an issue that relates to the use of logcheck over at #701 where @maxhq also suggested it's use along with some example configs. Seems like a great program but with regard to @MFlyer 's very recent log gui addition I think we should let it settle as is for a bit and maybe consider logcheck in the light of the other issue.

**Comment by henfri** [comment]
Hello Phillxnet,

yes, I am aware. 
I also sympathize with letting this feature (gui) settle. BUT: I am convinced that adding the filtering with the help of the logcheck filters to the GUI would *really* boost the usefullness.

As a user, I want to see important messages. Currently, they are hidden in a crowd of unimportant messages.

Greetings,
Hendrik

**Comment by MFlyer** [comment]
Thanks to @phillxnet for ref to other issue and to @henfri Hendrik for suggestion about logcheck/ logcheck rules integration in UI (will come back to this, that's a promise :wink: )

**Pull request merged**